test/suite/test_bug035.py

```diff
@@ -0,0 +1,107 @@
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import wiredtiger, wttest, threading
+# test_bug35.py
+# This python test uses control points to reproduce WT-12349.
+# 1. Populate data in the table
+# 2. Create tombstones on 50% of the table
+# 3. Force evict the tombstone data onto disk.
+# 4. Create two threads. One thread T1 will perform reads under isolation read-uncommitted and
+# aonther thread T2 will generate an update and a modify.
+# 5. Enable control point to control the execution of the concurrency between two threads.
+# 6. Pause T1 when the thread starts to reconstruct the modify in the update list.
+# 7. Once T2 generates an update, modify and finishes rollback, signal T1 to continue.
+# 8. T1 is now in an invalid state and should assert.
+class test_bug35(wttest.WiredTigerTestCase):
+    uri = 'table:test_bug035'
+    conn_config = 'cache_size=500MB,statistics=(all)'
+    nrows = 100
+
+    def evict_cursor(self, uri, nrows):
+        s = self.conn.open_session()
+        s.begin_transaction()
+        # Configure debug behavior on a cursor to evict the page positioned on when the reset API is used.
+        evict_cursor = s.open_cursor(uri, None, "debug=(release_evict)")
+        for i in range(1, nrows):
+            evict_cursor.set_key(i)
+            evict_cursor.search()
+            evict_cursor.reset()
+        s.rollback_transaction()
+        evict_cursor.close()
+
+    def construct_modify_upd_list(self):
+        session = self.setUpSessionOpen(self.conn)
+        cursor = session.open_cursor(self.uri)
+        session.begin_transaction("isolation=read-uncommitted")
+        for i in range(1, self.nrows):
+            cursor.set_key(i)
+            cursor.search()
+        session.commit_transaction()
+        cursor.close()
+
+        session.close()
+
+    def test_bug35(self):
+        create_params = 'key_format=r,value_format=S'
+        self.session.create(self.uri, create_params)
+        value1 = 'a' * 500
+        value2 = 'b' * 500
+
+        # Populate data in the data store table.
+        self.conn.set_timestamp('oldest_timestamp=' + self.timestamp_str(1))
+        cursor = self.session.open_cursor(self.uri)
+        self.session.begin_transaction()
+        for i in range(1, self.nrows):
+            cursor[i] = value1
+        self.session.commit_transaction('commit_timestamp=' + self.timestamp_str(5))
+
+        # Add in a tombstone.
+        self.session.begin_transaction()
+        for i in range(1, self.nrows // 2):
+            cursor.set_key(i)
+            self.assertEqual(cursor.remove(), 0)
+        self.session.commit_transaction('commit_timestamp=' + self.timestamp_str(6))
+        self.conn.set_timestamp('oldest_timestamp=' + self.timestamp_str(7))
+
+        self.evict_cursor(self.uri, self.nrows)
+        self.session.checkpoint()
+        self.conn.enable_control_point(6, "")
+        thread = threading.Thread(target=self.construct_modify_upd_list)
```

Comment:
It would be great to make it easier for readers of the code to match up the comments listing the steps and the code by, for example, renaming `thread` to `thread2`.

Reply:
Done, i have now laid out the steps inside the scenarios.

Reply:
Hi @jiechenbo , that's much easier to follow now - thank you.